ext/mbstring/libmbfl/filters/mbfilter_singlebyte.c

```diff
@@ -89,6 +98,48 @@ static int mbfl_conv_reverselookup_table(int c, mbfl_convert_filter *filter, int
 	static int mbfl_filt_conv_wchar_##id(int c, mbfl_convert_filter *filter) { \
 		return mbfl_conv_reverselookup_table(c, filter, tbl_min, tbl); \
 	} \
+	static size_t mb_##id##_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state) \
+	{ \
+		unsigned char *p = *in, *e = p + *in_len; \
+		uint32_t *out = buf, *limit = buf + bufsize; \
+		while (p < e && out < limit) { \
+			unsigned char c = *p++; \
+			*out++ = (c < tbl_min) ? c : coalesce(tbl[c - tbl_min], MBFL_BAD_INPUT); \
+		} \
+		*in_len = e - p; \
+		*in = p; \
+		return out - buf; \
+	} \
+	static mb_convert_buf* mb_wchar_to_##id(uint32_t *in, size_t len, mb_convert_buf *buf, bool end) \
+	{ \
+		buf = mb_convert_buf_ensure(buf, buf->len + len); \
+		while (len--) { \
+			uint32_t w = *in++; \
+			unsigned int error = 0; \
+	retry: \
+			if (w < tbl_min) { \
+				mb_convert_buf_add(buf, w & 0xFF); \
+			} else { \
+				for (int i = 0; i < 256 - tbl_min; i++) { \
+					if (w == tbl[i]) { \
+						mb_convert_buf_add(buf, i + tbl_min); \
+						goto continue_##id; \
+					} \
+				} \
+				if (MBSTRG(current_filter_illegal_mode) == MBFL_OUTPUTFILTER_ILLEGAL_MODE_CHAR) { \
```

Comment:
Hm, so this is the code that switches to `?` if the substitution character cannot be encoded in the target encoding? I don't really like the fact that this is being handled in the conversion function of specific encodings. Is there some way we can push this down into mb_illegal_output? Like, make it remember the current output position, try to print and if error increases reset and try again with a different char?

Reply:
Fixed.